### PR TITLE
[6326]Testy jednostkowe

### PR DESCRIPTION
**Pull request by efeniks**
**Ticket**: [BUG 6326](https://bugzilla.stxnext.pl/show_bug.cgi?id=6326)
Wykonano testy jednostkowe

### CODE REVIEW

File: src/presence_analyzer/views.py [hunk 1]
```diff
@@ -60,6 +60,7 @@ def presence_weekday_view(user_id):
     """
     Returns total presence time of given user grouped by weekday.
     """
+    # import ipdb; ipdb.set_trace()
```
Comment (mjelonek92) [inline src/presence_analyzer/views.py]: :wave:

File: src/presence_analyzer/tests.py [hunk 2]
```diff
@@ -52,6 +52,23 @@ def test_api_users(self):
         self.assertEqual(len(data), 2)
         self.assertDictEqual(data[0], {u'user_id': 10, u'name': u'User 10'})
 
+    def test_mean_time_weekday_view(self):
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: In this tests you only check if ```len(data)``` isn't equal to 0. I don't think so it is good test case. In ```data``` could be anything right now.

File: src/presence_analyzer/views.py [hunk 3]
```diff
@@ -71,5 +72,6 @@ def presence_weekday_view(user_id):
         for weekday, intervals in enumerate(weekdays)
     ]
 
-    result.insert(0, ('Weekday', 'Presence (s)'))
+    result.insert(0, 
+                 ('Weekday', 'Presence (s)'))
```
Comment (dpolcyn) [inline src/presence_analyzer/views.py]: Why did you add this code in to functionality. In bug description I don't see any information about editing it.

File: src/presence_analyzer/tests.py [hunk 4]
```diff
@@ -52,6 +52,23 @@ def test_api_users(self):
...
+        mean_time_data = views.mean_time_weekday_view(10)
+        self.assertEqual(mean_time_data.content_type, 'application/json')
+        data = json.loads(mean_time_data.data)
+        self.assertNotEqual(len(data), 0)
+        self.assertNotEqual(mean_time_data.status_code, 404)
+        self.assertEqual(mean_time_data.status_code, 200)
+
+    def test_presence_weekday_view(self):
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: Maybe it is worth to check if view return 404 if users doesn't exist?

File: src/presence_analyzer/tests.py [hunk 5]
```diff
@@ -52,6 +52,66 @@ def test_api_users(self):
         self.assertEqual(len(data), 2)
         self.assertDictEqual(data[0], {u'user_id': 10, u'name': u'User 10'})
 
+    def test_mean_time_weekday_view(self):
+        mean_time_url = '/api/v1/mean_time_weekday/10'
+        mean_time_data = self.client.get(mean_time_url)
+        self.assertEqual(mean_time_data.content_type, 'application/json')
+        self.assertEqual(mean_time_data.status, '200 OK')
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: Just check status code.

File: src/presence_analyzer/tests.py [hunk 6]
```diff
@@ -52,6 +52,66 @@ def test_api_users(self):
         self.assertEqual(len(data), 2)
         self.assertDictEqual(data[0], {u'user_id': 10, u'name': u'User 10'})
 
+    def test_mean_time_weekday_view(self):
+        mean_time_url = '/api/v1/mean_time_weekday/10'
+        mean_time_data = self.client.get(mean_time_url)
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: Why not just ```mean_time_data = self.client.get(/api/v1/mean_time_weekday/10)

File: src/presence_analyzer/tests.py [hunk 7]
```diff
@@ -52,6 +52,66 @@ def test_api_users(self):
...
+        self.assertEqual(mean_time_data.content_type, 'application/json')
+        self.assertEqual(mean_time_data.status, '200 OK')
+        data = json.loads(mean_time_data.data)
+        self.assertEqual(len(data), 7)
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: You still didn't changed it. Check my previous comment.

File: src/presence_analyzer/tests.py [hunk 8]
```diff
@@ -52,6 +52,66 @@ def test_api_users(self):
...
+        self.assertTrue(mean_time_data)
+        self.assertIsNotNone(mean_time_data)
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: This make no sense for me checking two times this same thing.

File: src/presence_analyzer/tests.py [hunk 9]
```diff
@@ -52,6 +52,66 @@ def test_api_users(self):
...
+        self.assertRegexpMatches(mean_time_url,
+                                 r'\/(\S+)\/(v\d)\/(\S+)\/(\d+)')
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: You don't have to check this url. You wrote it to tests by yourself.

File: src/presence_analyzer/tests.py [hunk 10]
```diff
@@ -52,6 +52,66 @@ def test_api_users(self):
...
+        self.assertNotIsInstance(mean_time_data, int)
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: Going this way you should check if it isn't instance of str, float etc... Don't do this.

File: src/presence_analyzer/tests.py [hunk 11]
```diff
@@ -52,6 +52,66 @@ def test_api_users(self):
...
+        self.assertEqual(mean_time_data.status_code, 200)
+        dataget = views.get_data()
+        self.assertIsInstance(dataget, dict)
+        self.assertIn(10, dataget.keys())
+        self.assertIn(datetime.date(2013, 9, 10), dataget[10].keys())
+        self.assertIn({'start': datetime.time(10, 48, 46),
+                       'end': datetime.time(17, 23, 51)}, dataget[10].values())
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: It is good to check whole dict.

**Comment by dpolcyn** [comment]
Squash commits.

File: src/presence_analyzer/tests.py [hunk 12]
```diff
@@ -52,6 +52,110 @@ def test_api_users(self):
         self.assertEqual(len(data), 2)
         self.assertDictEqual(data[0], {u'user_id': 10, u'name': u'User 10'})
 
+    def test_mean_time_weekday_view(self):
+        mean_time_data = self.client.get('/api/v1/mean_time_weekday/10')
+        self.assertEqual(mean_time_data.content_type, 'application/json')
+        data = json.loads(mean_time_data.data)
+        self.assertEqual(data,  [
+            [u'Mon', 0],
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: ```from __future__ import unicode_literals``` then you don't have to use ```u```

File: src/presence_analyzer/tests.py [hunk 13]
```diff
@@ -52,6 +52,110 @@ def test_api_users(self):
...
+            [u'Tue', 30047.0],
+            [u'Wed', 24465.0],
+            [u'Thu', 23705.0],
+            [u'Fri', 0],
+            [u'Sat', 0],
+            [u'Sun', 0],
+            ]
+        )
+        self.assertTrue(mean_time_data)
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: You already checked if something is there when u assert list.

File: src/presence_analyzer/tests.py [hunk 14]
```diff
@@ -52,6 +52,110 @@ def test_api_users(self):
...
+                'start': datetime.time(9, 39, 5),
+                'end': datetime.time(17, 59, 52)
+                    }
+                },
+            dataget[10])
+
+    def test_mean_time_weekday_view_404(self):
+        mean_time_data = self.client.get('/api/v1/mean_time_weekday/9')
+        self.assertEqual(mean_time_data.content_type, 'text/html')
+        self.assertTrue(mean_time_data)
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: Is there case when ```self.client.get('/api/v1/mean_time_weekday/9')``` return you False?

File: src/presence_analyzer/tests.py [hunk 15]
```diff
@@ -52,6 +52,110 @@ def test_api_users(self):
...
+        self.assertEqual(data, [
+            [u'Weekday', u'Presence (s)'],
+            [u'Mon', 0],
+            [u'Tue', 30047],
+            [u'Wed', 24465],
+            [u'Thu', 23705],
+            [u'Fri', 0],
+            [u'Sat', 0],
+            [u'Sun', 0]]
+        )
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: ```])```

File: src/presence_analyzer/tests.py [hunk 16]
```diff
@@ -85,6 +189,44 @@ def test_get_data(self):
...
+        self.assertIsInstance(inter_data, int)
+        self.assertIn(inter_data, [z for z in range(3500, 4500)])
+        self.assertLess(inter_data, 5000)
+        self.assertIsNotNone(inter_data)
+        self.assertNotAlmostEqual(inter_data, 1)
+
+    def test_mean(self):
+        mean_data = utils.mean([37435, 67])
+        self.assertIsInstance(mean_data, float)
+        self.assertLess(mean_data, 50000)
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: This assert mean it doesn't matter if this function return 1, 45 or 1999?

File: src/presence_analyzer/tests.py [hunk 17]
```diff
@@ -2,11 +2,13 @@
 """
 Presence analyzer unit tests.
 """
+from __future__ import unicode_literals
 import os.path
 import json
 import datetime
 import unittest
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: One line is enought.

File: src/presence_analyzer/tests.py [hunk 18]
```diff
@@ -2,11 +2,13 @@
 """
 Presence analyzer unit tests.
 """
+from __future__ import unicode_literals
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: Enter after this line.

File: src/presence_analyzer/tests.py [hunk 19]
```diff
@@ -85,6 +176,41 @@ def test_get_data(self):
...
+                {'end': datetime.time(17, 0, 42),
+                 'start': datetime.time(8, 19, 37)}}
+        )
+        self.assertIsInstance(group_data, list)
+        self.assertIsNotNone(group_data)
+        self.assertIn(group_data[1] or group_data[2] or group_data[3],
+                      [[x] for x in range(31265, 31280)])
+        self.assertEqual(
+            len(group_data), 7
+        )
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: It can be in one line.

File: src/presence_analyzer/tests.py [hunk 20]
```diff
@@ -85,6 +121,38 @@ def test_get_data(self):
...
+                                    datetime.time(17, 0, 42))
+        self.assertIsInstance(inter_data, int)
+        self.assertEqual(inter_data, 3602)
+        self.assertIn(inter_data, [z for z in range(3500, 4500)])
+        self.assertNotAlmostEqual(inter_data, 1)
+
+    def test_mean(self):
+        mean_data = utils.mean([37435, 67])
+        self.assertIsInstance(mean_data, float)
+        self.assertEqual(mean_data, 18751.0)
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: More cases.

File: src/presence_analyzer/tests.py [hunk 21]
```diff
@@ -85,6 +121,38 @@ def test_get_data(self):
...
+        self.assertIsInstance(group_data, list)
+        self.assertIn(group_data[1] or group_data[2] or group_data[3],
+                      [[x] for x in range(31265, 31280)])
+        self.assertEqual(len(group_data), 7)
+
+    def test_seconds_since_midnight(self):
+        mid_data = utils.seconds_since_midnight(datetime.time(17, 0, 42))
+        self.assertIsInstance(mid_data, int)
+        self.assertEqual(mid_data, 61242)
+        self.assertIn(mid_data, [y for y in range(60500, 61500)])
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: Again. Range? You know what should be there...

File: src/presence_analyzer/tests.py [hunk 22]
```diff
@@ -85,6 +121,68 @@ def test_get_data(self):
...
+    def test_seconds_since_midnight(self):
+        mid_data = utils.seconds_since_midnight(datetime.time(17, 0, 42))
+        self.assertIsInstance(mid_data, int)
+        self.assertEqual(mid_data, 61242)
+        assert mid_data != 0
+        mid_data_second = utils.seconds_since_midnight(
+                                                    datetime.time(16, 7, 41))
+        self.assertEqual(mid_data_second, 58061)
+        self.assertIsInstance(mid_data_second, int)
+        mid_data_third = utils.seconds_since_midnight(
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: I don't see any reason of creating variables like this: ```mid_data_third, mid_data_second```. After comparison you can override this variables.

File: src/presence_analyzer/tests.py [hunk 23]
```diff
@@ -50,7 +52,41 @@ def test_api_users(self):
...
+            ['Wed', 24465.0],
+            ['Thu', 23705.0],
+            ['Fri', 0],
+            ['Sat', 0],
+            ['Sun', 0],
+        ])
+        self.assertEqual(mean_time_data.status_code, 200)
+        dataget = views.get_data()
+        self.assertIsInstance(dataget, dict)
+        self.assertIn(10 or 11, dataget.keys())
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: Using ```mean_time_data = self.client.get('/api/v1/mean_time_weekday/10')``` accept if there is ```11``` in ```dataget.keys()```?

File: src/presence_analyzer/tests.py [hunk 24]
```diff
@@ -85,6 +129,59 @@ def test_get_data(self):
...
+        self.assertEqual(len(group_data), 7)
+
+    def test_seconds_since_midnight(self):
+        mid_data = utils.seconds_since_midnight(datetime.time(17, 0, 42))
+        self.assertIsInstance(mid_data, int)
+        self.assertEqual(mid_data, 61242)
+        assert mid_data != 0
+        mid_data = utils.seconds_since_midnight(datetime.time(16, 7, 41))
+        self.assertEqual(mid_data, 58061)
+        self.assertIsInstance(mid_data, int)
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: One time checking it is enough...

File: src/presence_analyzer/tests.py [hunk 25]
```diff
@@ -50,7 +52,49 @@ def test_api_users(self):
...
+            ['Weekday', 'Presence (s)'],
+            ['Mon', 0],
+            ['Tue', 30047],
+            ['Wed', 24465],
+            ['Thu', 23705],
+            ['Fri', 0],
+            ['Sat', 0],
+            ['Sun', 0]
+        ])
+        self.assertEqual(presence_data.status_code, 200)
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: Move it up

File: src/presence_analyzer/tests.py [hunk 26]
```diff
@@ -50,7 +52,49 @@ def test_api_users(self):
         self.assertEqual(resp.content_type, 'application/json')
         data = json.loads(resp.data)
         self.assertEqual(len(data), 2)
-        self.assertDictEqual(data[0], {u'user_id': 10, u'name': u'User 10'})
+        self.assertDictEqual(data[0], {'user_id': 10, 'name': 'User 10'})
+
+    def test_mean_time_weekday_view(self):
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: I don't see checking 404 status

File: src/presence_analyzer/tests.py [hunk 27]
```diff
@@ -85,6 +142,61 @@ def test_get_data(self):
...
+        )
+        self.assertEqual(inter_data, 31650)
+
+    def test_mean(self):
+        mean_data = utils.mean([37435, 67])
+        self.assertIsInstance(mean_data, float)
+        self.assertEqual(mean_data, 18751.0)
+        mean_data = utils.mean([29782, 839])
+        self.assertEqual(mean_data, 15310.5)
+        mean_data = utils.mean([27573, 29072])
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: More values in list?

File: src/presence_analyzer/tests.py [hunk 28]
```diff
@@ -50,7 +52,62 @@ def test_api_users(self):
...
+                'end': datetime.time(17, 59, 52)
+            }
+        )
+        self.assertIn(sample_dict, dataget[10].items())
+        sample_date = datetime.date(2013, 9, 10)
+        self.assertItemsEqual(
+            dataget[10][sample_date].keys(),
+            ['start', 'end']
+        )
+        self.assertItemsEqual(dataget.keys(), [10, 11])
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: It is already tested.

File: src/presence_analyzer/tests.py [hunk 29]
```diff
@@ -85,6 +142,78 @@ def test_get_data(self):
...
+            29992
+        ])
+        self.assertIsInstance(mean_data, float)
+        self.assertEqual(mean_data, 22963.25)
+        mean_data = utils.mean([
+            36271,
+            34577,
+            30658,
+            29401,
+            29655
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: Missing coma

File: src/presence_analyzer/tests.py [hunk 30]
```diff
@@ -85,6 +133,80 @@ def test_get_data(self):
             datetime.time(9, 39, 5)
         )
 
+    def test_group_by_weekday(self):
+        sample_data = utils.get_data()
+        group_data = utils.group_by_weekday(sample_data[10])
+        self.assertIsInstance(group_data, list)
```
Comment (mjelonek92) [inline src/presence_analyzer/tests.py]: I find most IsInstance asserts redundant if you use assertEqual. For lists or dicts use type specific asserts found here https://docs.python.org/2/library/unittest.html#unittest.TestCase.addTypeEqualityFunc
